staging/src/k8s.io/apimachinery/pkg/util/ip/family.go

```diff
@@ -0,0 +1,184 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ip
+
+import (
+	"net"
+	"net/netip"
+
+	netutils "k8s.io/utils/net"
+)
+
+// IPFamily represents the IP Family (IPv4 or IPv6) of an IP address or CIDR block.
+// (Note that this type is aliased as v1.IPFamily.)
+type IPFamily string
+
+const (
+	// IPv4Protocol indicates an IPv4 IP or CIDR. (This is the same value as
+	// v1.IPv4Protocol.)
+	IPv4Protocol IPFamily = "IPv4"
+	// IPv6Protocol indicates an IPv6 IP or CIDR. (This is the same value as
+	// v1.IPv6Protocol.)
+	IPv6Protocol IPFamily = "IPv6"
+	// IPFamilyUnknown indicates an unspecified or invalid IP family. (This is the
+	// same value as v1.IPFamilyUnknown.)
+	IPFamilyUnknown IPFamily = ""
+)
+
+type anyIPAddress interface {
+	net.IP | netip.Addr | string
+}
+
+type anyCIDRAddress interface {
+	*net.IPNet | netip.Prefix | string
+}
```

Comment:
is it worth this complexity?

Reply:
Yeah, I think the plan/hope is to converge on the `netip` types, but it's not going to happen overnight. Much of the golang stdlib still uses the `net` types (e.g. `crypto/x509` uses `net.IP` to refer to IP address altnames in certificates) and likewise things like `vishvananda/netlink` and `spf13/pflag` and `moby/ipvs` still use `net`.

And even ignoring the "both net and netip types" part, I found it nice to just have a single function for both parsed types and strings, rather than needing, eg, `IsIPv4(ip net.IP)` vs `IsIPv4String(ip string)` like netutils has.

(I even considered getting rid of the IP-vs-CIDR distinction and just having `IsIPv4()` that takes an IP, CIDR, or string, though that gets a bit more annoying to parse in the string case.)

Reply:
@aojea So how much do you hate the generics? :joy: 

I'm realizing that, as above, there are going to be a lot of places that still need to deal with the old IP/CIDR types for quite a while. And we don't want to force them all to do `utilip.IPFromAddr(utilip.ParseValidIP(ipstr))` because (a) that's very verbose, and (b) it doesn't actually work anyway because you're ignoring the `error` value returned from `ParseValidIP`.

But what if instead you could do `utilip.ParseValid[net.IP](ipstr)`? Which, of course, could reduce to just `utilip.ParseValid(ipstr)` in many contexts.

And then that pattern could be reused for all the other functions too. eg, if you have a unit test with an `expectedIP net.IP` field, you could just say `expectedIP: utilip.MustParse("1.2.3.4")` and the compiler can infer that you mean `utilip.MustParse[net.IP]("1.2.3.4")`.

(I think I'd still have `ParseIP`, `ParseValidIP`, `MustParseIP` etc as aliases for the `netip.Addr`-returning methods which we expect to _eventually_ make up the bulk of the usage.)

(I guess the alternative would be to have `ParseAddr`, `ParseValidAddr`, `MustParseAddr`, etc. returning `netip.Addr` vs `ParseIP`, `ParseValidIP`, `MustParseIP`, etc returning `net.IP`, and likewise with `Prefix` vs `CIDR`. But I don't really like that because I feel like it gives the "good" names to the "bad" functions...)